### PR TITLE
Feat/adp 2534 replace current fee calculation

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

We want to remove the dependency on the CML/CSL from the SDK, part of the functionality we are using from these libraries is related to TX min fee calculation. In particular two functions:

- Minimum Ada Required.
- Min Fee.

# Proposed Solution

Implement these two functions natively in the SDK.

# Important Changes Introduced

The minimum ADA required (per UTXO) and min fee is now calculated natively by the SDK.

### CODE REVIEW

File: packages/input-selection/src/selectionConstraints.ts [hunk 1]
```diff
@@ -26,44 +27,19 @@ export const computeMinimumCost =
...
-      // TODO: Get this from cardano-services
-      const priceMem = scope.manage(
-        CML.UnitInterval.new(scope.manage(CML.BigNum.from_str('577')), scope.manage(CML.BigNum.from_str('10000')))
-      );
-      const priceStep = scope.manage(
-        CML.UnitInterval.new(scope.manage(CML.BigNum.from_str('721')), scope.manage(CML.BigNum.from_str('100000000')))
-      );
-      const exInitPrices = scope.manage(CML.ExUnitPrices.new(priceMem, priceStep));
+    // TODO: Get this from cardano-services
+    const prices = { memory: 0.0577, steps: 0.000_007_21 };
```
Comment (mkazlauskas) [inline packages/input-selection/src/selectionConstraints.ts:34]: You can extend `ProtocolParametersRequiredByInputSelection` to get the `prices`

File: packages/input-selection/test/testData.ts [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+      referenceInputs: [txIn],
+      totalCollateral: 100n
+    },
+    id: Cardano.TransactionId('856c8bc6ce3725188b496d62fa389f2beff2f701e6d35af39d3f3464bbce0cec')
+  };
+
+  for (let i = 0; i < 10_000; ++i)
+    bigTx.body.referenceInputs!.push({
+      index: 0,
+      txId: Cardano.TransactionId(getRandomHash(64))
```
Comment (mkazlauskas) [inline packages/input-selection/test/testData.ts:273]: from `util-dev`
```suggestion
      txId: generateRandomHexString(64)
```

File: packages/input-selection/test/fees.test.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      }
+
+      output.value.assets = assets;
+      const minAda = minAdaRequired(output, COST_PER_UTXO_BYTE);
+      expect(minAda).toBe(160_599_220n);
+    });
+  });
+
+  describe('minFee', () => {
+    it('calculate the correct min fee for a babbage transaction', () => {
```
Comment (mkazlauskas) [inline packages/input-selection/test/fees.test.ts:113]:
```suggestion
    it('calculate the correct min fee for transaction without scripts', () => {
    it('calculate the correct min fee for transaction with scripts', () => {
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/input-selection/src/fees.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+/**
+ * Gets the minimum fee incurred by the scripts on the transaction.
+ *
+ * @param tx The transaction to compute the min script fee from.
+ * @param exUnitsPrice The prices of the execution units.
+ */
+const minScriptFee = (tx: Cardano.Tx, exUnitsPrice: Cardano.Prices): bigint => {
+  if (!tx.witness.redeemers) return BigInt(0);
+
+  const totalExUnits = getTotalExUnits(tx.witness.redeemers);
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/input-selection/src/fees.ts:61]: Thanks @AngelCastilloB

File: packages/input-selection/test/fees.test.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      const minAda = minAdaRequired(output, COST_PER_UTXO_BYTE);
+      expect(minAda).toBe(160_599_220n);
+    });
+  });
+
+  describe('minFee', () => {
+    it('calculate the correct min fee for transaction without scripts', () => {
+      const prices = { memory: 0.0577, steps: 0.000_007_21 };
+      const fee = minFee(babbageTxWithoutScript, prices, MinFeeConstant(155_381), MinFeeCoefficient(44));
+      expect(fee).toBe(206_905n);
```
Comment (mkazlauskas) [inline packages/input-selection/test/fees.test.ts:117]: Isn't this a little too high? I remember seeing fees at around .17

**Review by mkazlauskas — APPROVED** [review]
Nice work!

**Review by rhyslbw — COMMENTED** [review]
We have a bit of scope creep here introducing fee calculation. Thinking ahead to offering a library that handles lower level transaction construction, that `wallet` and our future `dapp` (name just a placeholder) package can depend on, how about one of the two options:

- rename `input-selection` to `tx-construction`
- add `tx-construction` with a dependency on `input-selection`, retaining the modularity we already have

File: packages/wallet/test/mocks/mockData.ts [hunk 6]
```diff
@@ -30,6 +30,7 @@ export const protocolParameters = {
   minFeeConstant: 155_381,
   minPoolCost: 340_000_000,
   poolDeposit: 500_000_000,
+  prices: { memory: 0.0577, steps: 0.000_007_21 },
```
Comment (mkazlauskas) [inline packages/wallet/test/mocks/mockData.ts:33]: Shouldn't this be in 7f6d706453bed26f318eb5a984c10712d444eef8?

**Review by mkazlauskas — DISMISSED** [review]
Nice work!

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Nice work @AngelCastilloB. Just a few spelling typo/mistakes in both commit messages to correct at the last squash

**Pull request merged**